feat: react component names post (#131)
getsentry/sentry.engineering: data/blog/improving-developer-experience-from-unreadable-css-selectors-to-clear-component-names.md

---

> title: 'Improving Developer Experience: From Unreadable CSS Selectors to Clear Component Names'

Reviewer: Small suggestion, to shorten the title a bit, and to explicitly call out `React` too:

```suggestion
title: 'Improving DX: From Unreadable CSS Selectors to Clear React Component Names'
```

date: '2024-05-30'

> tags: ['web', 'react']

Reviewer: ```suggestion
tags: ['web', 'react', 'debugging']
```

draft: false
summary: 'How to transform unreadable CSS selectors to React component names.'

> images: []

Author: missing image

layout: PostLayout

> authors: ['catherinelee']

Reviewer: Should @0Calories be added as a co-author? If so, his `.md` file also needs to be added

---

> Reading our own code is much easier than the output of a bundler or transpiler. The same goes for React developers reading CSS selectors generated from their CSS library of choice. CSS selectors are used all across Sentry to show clicks on components. This can provide useful information like which component is the cause of rage and dead clicks, and what component was clicked in a span. However, if you’re using React, these selectors become unreadable in production due to the minification process, so your selectors end up looking like this:

Reviewer: One might land on this post without knowing what `Sentry` is:

```suggestion
Reading our own code is much easier than the output of a bundler or transpiler. The same goes for React developers reading CSS selectors generated from their CSS library of choice. CSS selectors are used all across [Sentry](https://sentry.io/for/react/) to show clicks on components. This can provide useful information like which component is the cause of rage and dead clicks, and what component was clicked in a span. However, if you’re using React, these selectors become unreadable in production due to the minification process, so your selectors end up looking like this:
```

`button.en302zp1.app-191aavw.e16hd6vm2[role="button"][data-test-id="common-options"]`

> This makes it hard to search for specific clicks or spans, and to find that component in the codebase when using search. [We made this experience better](https://sentry.io/changelog/react-component-names-is-now-available/) by utilizing component names instead of CSS selectors, so you can easily view and search for components with one simple name, like this: `CommonOptions`. In order to do this, we first had to build on our [bundler plugin](https://www.npmjs.com/package/@sentry/bundler-plugin-core) to annotate output DOM with their component name when possible.

Reviewer: I wish we could link to a good resource that explains what `spans` are (outside the scope of Performance monitoring ideally). But I didn't find a good one. @drguthals might have ideas.

## Building the bundler plugin

> There are three main parts that make up the system of the component name feature: The Babel plugin, the bundler plugins, and the browser SDK. The component name plugin itself is a Babel plugin; it hooks into the transform step of your bundler’s build process, in order to attach React component names as additional properties on your compiled HTML. It parses the AST (Abstract Syntax Tree) of your JSX files at build-time to determine the names of your components, and attaches them as additional properties to the compiled HTML.

Reviewer: We're missing some links here as there's a lot of concepts and code:

```suggestion
There are three main parts that make up the system of the component name feature: The [Babel plugin](https://github.com/getsentry/sentry-javascript-bundler-plugins/tree/main/packages/babel-plugin-component-annotate), the [bundler plugins](https://github.com/getsentry/sentry-javascript-bundler-plugins/tree/main/packages/bundler-plugin-core), and the [Sentry React SDK](https://github.com/getsentry/sentry-javascript/tree/develop/packages/react). The component name plugin itself is a Babel plugin; it hooks into the transform step of your bundler’s build process, in order to attach React component names as additional properties on your compiled HTML. It parses the AST (Abstract Syntax Tree) of your JSX files at build-time to determine the names of your components, and attaches them as additional properties to the compiled HTML.
```
Also is it in the core browser SDK or React? I assumed React so suggested that ^

Reviewer 2: It's in the core browser SDK

> The bundler plugin itself is just a medium to easily get the React annotate plugin into your project, without needing to fiddle around with your Babel configuration. This of course also means that you can install the annotate plugin directly, without using our bundler plugins. However, our bundler plugins come packaged with other very useful features that will make setting up certain workflows on Sentry a breeze, so we recommend you use them if possible!

Reviewer: > that will make setting up certain workflows on Sentry a breeze

Can we mention _for example A and B_? or link to somewhere that explains what workflows that helps with?

> The browser SDK is what is installed in your application’s frontend to capture events and send them to Sentry in the form of issues, spans, and replays. As of version 7.91.0, the Sentry JavaScript SDK parses your application’s HTML for the properties that were added by the plugin, and attaches them as complimentary data to your events.

Reviewer: Is it browser SDK (`@sentry/browser` or Sentry SDK for React `@sentry/react`)?

---
title: 'Improving Developer Experience: From Unreadable CSS Selectors to Clear Component Names'
date: '2024-05-30'
tags: ['web', 'react']
draft: false
summary: 'How to transform unreadable CSS selectors to React component names.'
images:
  [
    /images/improving-developer-experience-from-unreadable-css-selectors-to-clear-component-names/hero.jpg,
  ]
layout: PostLayout

> authors: ['catherinelee']

Author: @0Calories add your name here after you create a file in authors

The browser SDK is what is installed in your application’s frontend to capture events and send them to Sentry in the form of issues, spans, and replays. As of version 7.91.0, the Sentry JavaScript SDK parses your application’s HTML for the properties that were added by the plugin, and attaches them as complimentary data to your events.

One of our core values at Sentry is “For Every Developer”. In order to stay true to this, we built this feature so that it could be used with as many bundlers as possible, and with ease of access in mind so that it could be installed and set up quickly. On release, this feature works out of the box for developers using our bundler plugins for Vite, Webpack, and Rollup. We used [Unplugin](https://unplugin.unjs.io/guide/) to build this, which standardizes plugin development across multiple bundlers into one single API.

## Adding React component names into Sentry

> There were many opportunities across the product where we could leverage component names: spans, breadcrumbs, rage click issues, and [dead and rage click selectors](https://blog.sentry.io/introducing-rage-and-dead-click-detection-for-session-replay/).

Reviewer: ```suggestion
There were many opportunities across the product where we could leverage component names: spans, breadcrumbs, [rage click issues, and dead and rage click selectors](https://blog.sentry.io/introducing-rage-and-dead-click-detection-for-session-replay/).
```

> ![React component name used in spans](/images/improving-developer-experience-from-unreadable-css-selectors-to-clear-component-names/span.png)

Reviewer: ```suggestion
### Spans, in Tracing and Performance Monitoring

![React component name used in spans](/images/improving-developer-experience-from-unreadable-css-selectors-to-clear-component-names/span.png)
```

_In spans, React component names are used in click interactions._

> ![React component name used in replay breadcrumbs](/images/improving-developer-experience-from-unreadable-css-selectors-to-clear-component-names/replay-breadcrumb.png)

Reviewer: ```suggestion
### Session Replay Breadcrumbs

![React component name used in replay breadcrumbs](/images/improving-developer-experience-from-unreadable-css-selectors-to-clear-component-names/replay-breadcrumb.png)
```

_In Replay breadcrumbs, React component names are used in the selector path in user click breadcrumbs. If the user clicked on a component with a component name, you can search replays by that component name._

> ![React component name used in rage and dead click selectors](/images/improving-developer-experience-from-unreadable-css-selectors-to-clear-component-names/rage-and-dead-selector.png)

Reviewer: ```suggestion
### Session Replay Index Page - Rage and Dead Click Cards

![React component name used in rage and dead click selectors](/images/improving-developer-experience-from-unreadable-css-selectors-to-clear-component-names/rage-and-dead-selector.png)
```

_In rage and dead click selectors, React component names are used as part of the selector name._

> ![React component name used in rage click issues](/images/improving-developer-experience-from-unreadable-css-selectors-to-clear-component-names/rage-click-issue.png)

Reviewer: ```suggestion
### Rage Click Issue Details

![React component name used in rage click issues](/images/improving-developer-experience-from-unreadable-css-selectors-to-clear-component-names/rage-click-issue.png)
```

_In rage click issues, the React component name is also shown if it’s available._

> The SDK had to be updated first to take in these new attributes. This involved creating a function to extract the component name from the attributes and propagating it throughout spans, UI breadcrumbs, and Replay breadcrumbs. Next, we updated Sentry’s UI to utilize these component names. In most areas, the component names just replaced the selector names, but for rage click issues, and dead and rage click selectors this was not the case. For rage click issues, the Clicked Element uses the selector name which provides useful context about the element. Hence decided to add an additional field for the component name: React Component Name. For rage and dead clicks, additional context besides the React component names are useful for debugging – for instance, `aria-label` and `data-test-id`. For that reason we decided to keep the selectors names. Since class names become minified, we decided to replace them with the react component name. Hence, rage and dead click selectors are now a combination of the component name and the additional context.

Reviewer: ```suggestion
The SDK had to be updated first to take in these new attributes. This involved creating a function to extract the component name from the attributes and propagating it throughout spans, UI breadcrumbs, and Replay breadcrumbs. Next, we updated Sentry’s UI to utilize these component names. In most areas, the component names just replaced the selector names, but for rage click issues, and dead and rage click selectors this was not the case. For rage click issues, the Clicked Element uses the selector name which provides useful context about the element. Hence we decided to add an additional field for the component name: React Component Name. For rage and dead clicks, additional context besides the React component names are useful for debugging – for instance, `aria-label` and `data-test-id`. For that reason we decided to keep the selectors names. Since class names become minified, we decided to replace them with the react component name and because of that, rage and dead click selectors are now a combination of the component name and the additional context.
```
